api/tests/devcomp/unit/infrastructure/datasources/learning-content/module-datasource_test.js

```diff
@@ -1,146 +1,70 @@
-                  },
-                  id: '9bd65c37-ea5d-49a9-a59c-4e07ab3f1049',
-                  instruction: '<p>Toutes les adresses mails se terminent par gmail.com ?</p>',
-                  proposals: [
-                    {
-                      content: 'Vrai',
-                      id: '7ebabab5-6acd-4660-bf33-5cc7206e7288',
-                    },
-                    {
-                      content: 'Faux',
-                      id: 'f56bb715-714e-4a94-ba34-eef5ccd755d0',
-                    },
-                  ],
-                  solution: 'f56bb715-714e-4a94-ba34-eef5ccd755d0',
-                  type: 'qcu',
-                },
-              ],
-              id: '71792e45-966a-4070-b4fc-9e8a773c3f6f',
-              title: 'Diversité des identifiants et des fournisseurs d’adresse mail',
-              type: 'activity',
-            },
-          ],
-          id: 'f7b3a2e1-0d5c-4c6c-9c4d-1a3d8f7e9f5d',
-          slug: 'bien-ecrire-son-adresse-mail',
-          title: 'Bien écrire son adresse mail',
-        };
-        expect(module).to.deep.equal(expectedModuleData);
+        const textElementSchema = Joi.object({
+          id: Joi.string()
+            .regex(/^[a-z0-9-]+$/)
+            .required(),
+          type: Joi.string().valid('text').required(),
+          content: Joi.string().required(),
+        }).required();
+
+        const qcuElementSchema = Joi.object({
+          id: Joi.string()
+            .regex(/^[a-z0-9-]+$/)
+            .required(),
+          type: Joi.string().valid('qcu').required(),
+          instruction: Joi.string().required(),
+          proposals: Joi.array()
+            .items({
+              id: Joi.string()
+                .regex(/^[a-z0-9-]+$/)
+                .required(),
+              content: Joi.string().required(),
+            })
+            .required(),
+          feedbacks: Joi.object({
+            invalid: Joi.string().required(),
+            valid: Joi.string().required(),
+          }).required(),
+          solution: Joi.string()
+            .regex(/^[a-z0-9-]+$/)
+            .required(),
+        }).required();
+
+        const moduleSchema = Joi.object({
+          id: Joi.string().guid().required(),
+          slug: Joi.string()
+            .regex(/^[a-z0-9-]+$/)
+            .required(),
+          title: Joi.string().required(),
+          grains: Joi.array()
+            .items(
+              Joi.object({
+                id: Joi.string().guid().required(),
+                type: Joi.string().valid('lesson', 'activity').required(),
+                title: Joi.string().required(),
+                elements: Joi.array().items(Joi.alternatives().try(textElementSchema, qcuElementSchema)).required(),
+              }).required(),
+            )
+            .required(),
+        }).required();
+
+        // When
+        const result = moduleSchema.validate(module);
+
+        // Then
+        expect(result.error).to.equal(undefined);
```

Comment:
L'erreur est un peu imbuvable quand ça pète, à voir comment améliorer les logs un jour

```
AssertionError: expected ValidationError: "grains[0].id" must be a… { …(2) } to equal undefined
```

Reply:
Je sortirai bien moduleschema, uuidschema etc. du describe pour les rendre globaux et simplifier la lecture du test en lui-même

Reply:
Tu devrais pouvoir faire quelque chose comme
```suggestion
        expect(result.error).to.equal(undefined, 'Mon super message d'erreur quand ça casse');
```

Reply:
Done